File: frontend/src/utils/nameExtractor.js
export const deriveClientName = (lead) => {
  if (!lead) return { fullName: 'N/A', firstName: 'N/A', lastName: 'N/A' };

  // If lead already has valid full_name and first_name set
  if (lead.full_name && lead.full_name !== 'N/A' && lead.full_name !== 'Not Available') {
    const parts = lead.full_name.trim().split(/\s+/);
    return {
      fullName: lead.full_name,
      firstName: lead.first_name && lead.first_name !== 'N/A' ? lead.first_name : (parts[0] || 'N/A'),
      lastName: lead.last_name && lead.last_name !== 'N/A' ? lead.last_name : (parts.length > 1 ? parts.slice(1).join(' ') : '')
    };
  }

  const company = lead.company_name || '';
  const email = lead.email || '';

  // 1. Try email extraction if available and non-generic
  if (email && email.includes('@') && email !== 'Not Available') {
    const local = email.split('@')[0].toLowerCase();
    const genericPrefixes = ['info', 'contact', 'admin', 'support', 'sales', 'hello', 'help', 'office', 'service', 'mail', 'inquiry', 'billing', 'team'];
    if (!genericPrefixes.includes(local) && local.length >= 3) {
      const clean = local.replace(/[._-]+/g, ' ').replace(/\d+/g, '').trim();
      const words = clean.split(/\s+/).filter(w => w.length > 1).map(w => w.charAt(0).toUpperCase() + w.slice(1));
      if (words.length >= 2) {
        return {
          fullName: words.join(' '),
          firstName: words[0],
          lastName: words.slice(1).join(' ')
        };
      } else if (words.length === 1) {
        return {
          fullName: `${words[0]} Client`,
          firstName: words[0],
          lastName: 'Client'
        };
      }
    }
  }

  if (!company || company === 'Not Available') {
    return { fullName: 'Client Lead', firstName: 'Client', lastName: 'Lead' };
  }

  // 2. Clean company name
  let cleaned = company.trim();

  // Strip common prefixes
  cleaned = cleaned.replace(/^(the\s+)?(law\s+offices?|offices?|law\s+firm|office|dr\.?|doctor|attorney|advocate)\s+of\s+/i, '');
  cleaned = cleaned.replace(/^(the\s+)?(law\s+offices?|offices?|law\s+firm|office|dr\.?|doctor|attorney|advocate)\s+/i, '');

  // Strip common suffixes
  cleaned = cleaned.replace(/,?\s*(p\.?c\.?|l\.?l\.?p\.?|l\.?l\.?c\.?|inc\.?|corp(oration)?|ltd(imited)?)$/i, '');
  cleaned = cleaned.replace(/\s+(law\s+firm|law\s+offices?|legal\s+group|legal\s+services|law\s+associates|associates|&amp;\s+associates|&\s+associates|and\s+associates|partners|&amp;\s+partners|&\s+partners|and\s+partners|group|practice|clinic|center|centre).*/i, '');
  cleaned = cleaned.replace(/\s*-\s*.*$/i, ''); // e.g. - New York

  cleaned = cleaned.trim();
  if (!cleaned) cleaned = company.trim();

  const words = cleaned.split(/\s+/).filter(Boolean);

  if (cleaned.includes('&') || cleaned.toLowerCase().includes('and')) {
    const firstWord = words[0] || cleaned;
    const lastWord = words[words.length - 1] !== firstWord ? words[words.length - 1] : 'Partner';
    return {
      fullName: cleaned,
      firstName: firstWord,
      lastName: lastWord
    };
  } else if (words.length >= 2) {
    const first = words[0].charAt(0).toUpperCase() + words[0].slice(1);
    const last = words.slice(1).map(w => w.charAt(0).toUpperCase() + w.slice(1)).join(' ');
    return {
      fullName: `${first} ${last}`,
      firstName: first,
      lastName: last
    };
  } else if (words.length === 1) {
    const first = words[0].charAt(0).toUpperCase() + words[0].slice(1);
    return {
      fullName: `${first} Representative`,
      firstName: first,
      lastName: 'Representative'
    };
  }

  return {
    fullName: company,
    firstName: company.split(' ')[0],
    lastName: 'Client'
  };
};

File: backend/app/services/name_extractor.py
import re
from typing import Dict, Optional, Tuple

PREFIX_PATTERNS = [
    r"^the\s+law\s+offices?\s+of\s+",
    r"^law\s+offices?\s+of\s+",
    r"^offices?\s+of\s+",
    r"^the\s+law\s+firm\s+of\s+",
    r"^law\s+firm\s+of\s+",
    r"^the\s+office\s+of\s+",
    r"^office\s+of\s+",
    r"^dr\.?\s+",
    r"^doctor\s+",
    r"^attorney\s+",
    r"^advocate\s+",
]

SUFFIX_PATTERNS = [
    r"\s*,?\s*p\.?c\.?$",
    r"\s*,?\s*l\.?l\.?p\.?$",
    r"\s*,?\s*l\.?l\.?c\.?$",
    r"\s*,?\s*inc\.?$",
    r"\s*,?\s*corp\.?(oration)?$",
    r"\s*,?\s*ltd\.?(ited)?$",
    r"\s+law\s+firm.*$",
    r"\s+law\s+offices?.*$",
    r"\s+legal\s+group.*$",
    r"\s+legal\s+services.*$",
    r"\s+law\s+associates.*$",
    r"\s+associates.*$",
    r"\s+&amp;\s+associates.*$",
    r"\s+&\s+associates.*$",
    r"\s+and\s+associates.*$",
    r"\s+partners.*$",
    r"\s+&amp;\s+partners.*$",
    r"\s+&\s+partners.*$",
    r"\s+and\s+partners.*$",
    r"\s+group$",
    r"\s+practice$",
    r"\s+clinic$",
    r"\s+center$",
    r"\s+centre$",
    r"\s*-\s*.*$",
]

GENERIC_EMAIL_PREFIXES = {
    "info", "contact", "admin", "support", "sales", "hello", "help", "office",
    "service", "mail", "inquiry", "inquiries", "billing", "general", "team"
}


def extract_name_from_email(email: Optional[str]) -> Optional[Tuple[str, str, str]]:
    if not email or "@" not in email or email == "Not Available":
        return None
    local = email.split("@")[0].lower().strip()
    if local in GENERIC_EMAIL_PREFIXES or len(local) < 3:
        return None

    clean = re.sub(r"[._-]+", " ", local)
    clean = re.sub(r"\d+", "", clean).strip()
    words = [w.capitalize() for w in clean.split() if len(w) > 1]

    if len(words) == 1:
        first = words[0]
        return f"{first} Client", first, "Client"
    elif len(words) >= 2:
        first = words[0]
        last = " ".join(words[1:])
        return f"{first} {last}", first, last
    return None


def derive_client_name(
    company_name: str,
    email: Optional[str] = None,
    existing_full: Optional[str] = None,
    existing_first: Optional[str] = None,
    existing_last: Optional[str] = None
) -> Dict[str, str]:
    """
    Derives appropriate client/contact names for business leads so N/A is never displayed.
    """
    if existing_full and existing_full not in ("N/A", "Not Available"):
        first = existing_first or existing_full.split()[0]
        last = existing_last or (" ".join(existing_full.split()[1:]) if len(existing_full.split()) > 1 else "")
        return {"full_name": existing_full, "first_name": first, "last_name": last}

    # 1. Try email extraction if non-generic
    email_res = extract_name_from_email(email)
    if email_res:
        return {"full_name": email_res[0], "first_name": email_res[1], "last_name": email_res[2]}

    # 2. Extract from company_name
    if not company_name or company_name in ("Not Available", "N/A"):
        return {"full_name": "Client Lead", "first_name": "Client", "last_name": "Lead"}

    cleaned = company_name.strip()

    # Strip prefixes
    for pat in PREFIX_PATTERNS:
        cleaned = re.sub(pat, "", cleaned, flags=re.IGNORECASE).strip()

    # Strip suffixes
    for pat in SUFFIX_PATTERNS:
        cleaned = re.sub(pat, "", cleaned, flags=re.IGNORECASE).strip()

    if not cleaned:
        cleaned = company_name.strip()

    parts = [p.capitalize() for p in cleaned.split() if p.strip()]

    if "&" in cleaned or "and" in [p.lower() for p in parts]:
        full = cleaned
        first = parts[0] if parts else cleaned
        last = parts[-1] if len(parts) > 1 else "Partner"
        return {"full_name": full, "first_name": first, "last_name": last}
    elif len(parts) >= 2:
        first = parts[0]
        last = " ".join(parts[1:])
        full = f"{first} {last}"
        return {"full_name": full, "first_name": first, "last_name": last}
    elif len(parts) == 1:
        first = parts[0]
        last = "Representative"
        full = f"{first} Representative"
        return {"full_name": full, "first_name": first, "last_name": last}

    return {"full_name": company_name, "first_name": company_name.split()[0], "last_name": "Client"}
